python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -1223,8 +1245,461 @@ def asset_check_spec(self) -> AssetCheckSpec:
+    "get_output_metadata": "use op_execution_context.op_def.get_output(...).metadata",
+    "output_for_asset_key": INPUT_OUTPUT_ALT,
+    "selected_output_names": INPUT_OUTPUT_ALT,
+}
+
+ALTERNATE_AVAILABLE_METHODS = {
+    "has_tag": "use dagster_run.has_tag instead",
+    "get_tag": "use dagster_run.get_tag instead",
+    "run_tags": "use dagster_run.tags instead",
+    "set_data_version": "use MaterializeResult instead",
+    "run": "use dagster_run instead.",
+}
+
+# TODO - add AssetCheck related methods to this list
+
+
+def _get_deprecation_kwargs(attr: str):
+    deprecation_kwargs = {"breaking_version": "1.7.0"}
+    deprecation_kwargs["subject"] = f"AssetExecutionContext.{attr}"
+
+    if attr in OP_EXECUTION_CONTEXT_ONLY_METHODS:
+        deprecation_kwargs["additional_warn_text"] = (
+            f"You have called the deprecated method {attr} on AssetExecutionContext. Use"
+            " the underlying OpExecutionContext instead by calling"
+            f" context.op_execution_context.{attr}."
+        )
+
+    if attr in DEPRECATED_IO_MANAGER_CENTRIC_CONTEXT_METHODS:
+        alt = DEPRECATED_IO_MANAGER_CENTRIC_CONTEXT_METHODS[attr]
+        deprecation_kwargs["additional_warn_text"] = (
+            f"You have called method {attr} on AssetExecutionContext that is oriented"
+            f" around I/O managers. If you not using I/O managers we suggest you {alt}. If"
+            " you are using I/O managers the method still exists at"
+            f" context.op_execution_context.{attr}."
+        )
+
+    if attr in ALTERNATE_AVAILABLE_METHODS:
+        deprecation_kwargs["additional_warn_text"] = f"Instead {ALTERNATE_AVAILABLE_METHODS[attr]}."
+
+    return deprecation_kwargs
+
+
+class AssetExecutionContext(OpExecutionContext):
+    def __init__(self, op_execution_context: OpExecutionContext) -> None:
+        self._op_execution_context = check.inst_param(
+            op_execution_context, "op_execution_context", OpExecutionContext
+        )
+
+    @public
+    @property
+    def op_execution_context(self) -> OpExecutionContext:
+        return self._op_execution_context
+
+    # IContext interface methods
+
+    @property
+    def is_asset_step(self) -> bool:
+        return self.op_execution_context.has_assets_def
+
+    @public
+    @property
+    def asset_key(self) -> AssetKey:
+        return self._op_execution_context.asset_key
+
+    @property
+    def asset_keys(self) -> Sequence[AssetKey]:
+        return list(self.op_execution_context.assets_def.keys_by_output_name.values())
+
+    @property
+    def provenance(self) -> Optional[DataProvenance]:
+        return self.get_asset_provenance(self.asset_key)
+
+    @property
+    def provenance_by_asset_key(self) -> Mapping[AssetKey, Optional[DataProvenance]]:
+        provenance_map = {}
+        for key in self.asset_keys:
+            provenance_map[key] = self.get_asset_provenance(key)
+
+        return provenance_map
```

Comment:
while I think it is defensible to have separate `asset_key` and `asset_keys` properties for convenience, for the remainder of these properties we should not have both variants.

My recommendation is to only allow lookup by key, and also to strongly consider a single object.

Reply:
that's fine w me - we'll also need to update ExtContext to match

Reply:
>while I think it is defensible to have separate asset_key and asset_keys properties for convenience, for the remainder of these properties we should not have both variants.
>
>My recommendation is to only allow lookup by key, and also to strongly consider a single object.

IMO this makes for annoying UX:

```
context.provenance_by_asset_key(context.asset_key)

# instead of

context.provenance
```

I think it should be:

```
def provenance(self, asset_key: Optional[AssetKey]):
```

Then if there is only a single asset you default to that, otherwise require the `asset_key`.

Reply:
exploring a bit more on the single object thread. if i understand correctly, the object could be some kind of `AssetInfo` class that would hold the `code_version`, `provenance`, and maybe things like `partition_keys` and `partition_time_window`. Rather than having a set of `code_version_by_asset_key`, `partition_time_window_by_asset_key` etc methods, we'd just have a `asset_info_by_asset_key` that would return these objects. 

it feels similar to `AssetSpec` but bound to a particular materialization, not a general definition

Reply:
> it feels similar to `AssetSpec` but bound to a particular materialization, not a general definition

Great mental model.

How about a compromise between the two approaches, asI do find @smackesey's concern of divergence between single and multi asset case convincing:

```python 

def asset_info(self, asset_key: Optional[AssetKey] = None) -> AssetInfo
   ...
```

However I do not think we should name this `AssetInfo`. @jamiedemaria what is the complete set of properties that will exist on this object? That might help with naming.

`partition_keys` and `partition_time_window` are also interesting. They only vary on a per-asset-key basis when partition mappings are present, so if someone doesn't know about mappings they may find a "by_key" variant of this APi confusing.

Reply:
Maybe not fully complete, but the properties I can think to include are:

* `code_version`
* `provenance`
* `partition_time_window`
* `partition_key_range`
* `partition_keys`
* `upstream_result_value` - no good name for this right now, but I'm thinking if we allow this:
```python
@asset 
def my_asset():
    return MaterializeResult(value="foo")
```
then this object would be a good way to give that value to the user downstream 
```python
@asset(
    deps=[my_asset]
)
def another_asset(context):
    context.asset_info_by_asset_key[my_asset].upstream_result_value == "foo"
```
* metadata
* potentially even things like the time the asset was materialized. There's a small collection of user questions in the "how to i get X info about the upstream asset?" category. we could look through those and see if there's anything worth including